sysvabi64/sysvabi64.rst

```diff
@@ -1844,6 +1854,601 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+a single TLS variable.
+
+The code sequence for local dynamic is the same as global dynamic and
+like global dynamic must be emitted exactly as specified. There are no
+specific relocations for Local Dynamic using the descriptor dialect. A
+special symbol ``_TLS_MODULE_BASE_`` is used to get a tlsdesccall to
+return the offset of the module's TLS block from the thread pointer.
+
+Code-generators are not required to implement local dynamic and can
+emit general dynamic in its place.
+
+Initial Exec
+^^^^^^^^^^^^
+
+Initial Exec can be used for static TLS. The location of the module's
+TLS block and the offset of the TLS variable within that block are
+run-time constants. The dynamic-loader computes the offset from the
+thread-pointer and places it in a GOT entry. The GOT entry is
+relocated by dynamic relocation ``R_AARCH64_TLS_TPREL64``.
+
+A shared-library that contains Initial Exec TLS must have the
+``DF_STATIC_TLS`` dynamic tag set. An attempt to load a shared library
+with ``DF_STATIC_TLS`` via ``dlopen`` will be rejected.
+
+Small Code model
+
+The static linker is permitted to relax the instructions below to
+Local Exec individually using the relocation directive. The
+instructions do not have to be contiguous.
+
+.. code-block:: asm
+
+    adrp xn, :gottprel: var            // R_AARCH64_TLSIE_ADR_GOTTPREL_PAGE21 var
+    ldr  xn, [xn, #:gottprel_lo12:var] // R_AARCH64_TLSIE_LD64_GOTTPREL_LO12_NC var
+    // offset of var from tp in xn
+
+Tiny Code model
+
+.. code-block:: asm
+
+    ldr  xn, :gottprel:var // R_AARCH64_TLSIE_LD_GOTTPREL_PREL19 var
+    // offset of var from tp in xn
+
+Large Code model
+
+.. code-block:: asm
+
+    movz xn, #:gottprel_g1:var    // R_AARCH64_TLSIE_MOVW_GOTTPREL_G1 var
+    movk xn, #:gottprel_g0_nc:var // R_AARCH64_TLSIE_MOVW_GOTTPREL_G0_NC var
+    ldr xn, [gp, xn]
+    // offset of var from tp in xn
+
+Local Exec
+^^^^^^^^^^
+
+Local Exec is used for accesses to the executable's TLS block. The
+executable always has the TLS module index of 1 so the offsets of the
+TLS variables from the thread pointer are static link time
+constants. The code sequences are the same for all code models.
+
+The instruction sequences below are not ABI. Using the instructions
+and relocations below increases the chances of static linkers applying
+the relaxations in (AAELF64_) when the size of the executables TLS
+block is smaller than 16 KiB.
+
+.. code-block:: asm
+
+    add  xn, xn, :tprel_hi12:var, lsl #12 // R_AARCH64_TLSLE_ADD_TPREL_HI12 var
+    add  xn, xn, :tprel_lo12_nc:var       // R_AARCH64_TLSLE_ADD_TPREL_LO12_NC var
+    // offset of var from tp in xn
+
+Optimization to load a 64-bit var directly into a core register.
+
+.. code-block:: asm
+
+    add  xn, tp, :tprel_hi12:var, lsl #12 // R_AARCH64_TLSLE_ADD_TPREL_HI12 var
+    ldr  xn, [xn, #:tprel_lo12_nc:var]    // R_AARCH64_TLSLE_LDST64_TPREL_LO12_NC var
+
+Static link time TLS Relaxations
```

Comment:
Perhaps call this Optimization to be consistent with x86/ppc and "Relocation optimization" (ADRP) and leave the term "relocation relaxation"  for RISC-V style section shrinking.